### PR TITLE
Supporting arbitrary number of leading dimensions

### PR DESCRIPTION
**Pull request by justanhduc**
### Description
This is the PR for #1068.
These first commits include:

1. Added `perform_keep_shape_video` decorator. Utilizing `_to_bchw` and `_to_bcdhw` in the decorator.
2. Fixed some bugs for functions supposed to support any number of leading dimensions, but not due to a hardcode bug.
3. Added support for the `kornia.enhance.adjust`.

Further, before moving on, I would like to discuss the following matters

1. How can I write tests for arbitrary dimensions? Could you suggest some sample?
2. JIT tests break due to `*args` and `**kwargs` in the decorators (currently I skip these tests).
3. Some users' codes may break when inputs not have shape `(b, c, h, w)` because some functions use `_to_bchw` and `_to_bcdhw`.

Please let me know your comments. Thanks!

### Status
**Work in progress**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>

  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?

</details>

### CODE REVIEW

**Review by edgarriba — APPROVED** [review]
LGTM. Please, rebase from master.

**Comment by justanhduc** [comment]
@edgarriba should I convert all the modules and merge when the PR is finished or convert and merge module by module? Currently this PR contains the conversion of only the enhance module.

**Comment by edgarriba** [comment]
@justanhduc to do it more smooth, I'd make a separated PR per module or groups or modules

**Comment by justanhduc** [comment]
@edgarriba Ok cool! Let me rebase and push.

**Comment by justanhduc** [comment]
@edgarriba I did something stupid and had to force push to undo. Sorry for the mess. Plz let me know if it is a problem.
Btw there is a failed test on Windows with python 3.6. Seems like the error comes from the Conda server, but I'm not sure.

File: kornia/enhance/adjust.py [hunk 1]
```diff
@@ -126,7 +126,7 @@ def adjust_hue_raw(input: torch.Tensor, hue_factor: Union[float, torch.Tensor])
     # if ((hue_factor < -pi) | (hue_factor > pi)).any():
     #     raise ValueError(f"Hue-factor must be in the range [-PI, PI]. Got {hue_factor}")
 
-    for _ in input.shape[1:]:
+    for _ in input.shape[-3:]:
```
Comment (edgarriba) [inline kornia/enhance/adjust.py:129]:
```suggestion
    for _ in range(len(input.shape) - len(saturation_factor.shape))
```

Comment (edgarriba) [reply kornia/enhance/adjust.py:129]: same as https://github.com/kornia/kornia/pull/1113/files#diff-ff1ae62fe6ba5559b9ad03f2ad9006d1c42c4b035fb7016a861c27251a523df8R241

or eventually having a `unsqueeze_as(input, data, dim=-1)`

File: kornia/enhance/adjust.py [hunk 2]
```diff
@@ -192,14 +192,14 @@ def adjust_gamma(
     The input image is expected to be in the range of [0, 1].
 
     Args:
-        input: Image to be adjusted in the shape of :math:`(*, N)`.
+        input: Image to be adjusted in the shape of :math:`(*, H, W)`.
```
Comment (edgarriba) [inline kornia/enhance/adjust.py:195]: what's the final decision about shape ? I see some HxW vs CxHxW - we should state this somewhere in the documentation. /cc @shijianjian @ducha-aiki @lferraz

File: kornia/utils/image.py [hunk 3]
```diff
@@ -65,23 +64,25 @@ def _to_bchw(tensor: torch.Tensor) -> torch.Tensor:
     if len(tensor.shape) == 3:
         tensor = tensor.unsqueeze(0)
 
+    if len(tensor.shape) > 4:
+        tensor = tensor.view(-1, tensor.shape[-3], tensor.shape[-2], tensor.shape[-1])
+
     return tensor
 
 
 def _to_bcdhw(tensor: torch.Tensor) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/utils/image.py:73]:
```suggestion
def tensor_to_bcdhw(tensor: torch.Tensor) -> torch.Tensor:
```
@shijianjian @ducha-aiki @lferraz to improve functional naming ?

File: kornia/utils/image.py [hunk 4]
```diff
@@ -47,16 +47,15 @@ def _to_bchw(tensor: torch.Tensor) -> torch.Tensor:
     """Converts a PyTorch tensor image to BCHW format.
 
     Args:
-        tensor (torch.Tensor): image of the form :math:`(H, W)`, :math:`(C, H, W)`, :math:`(H, W, C)` or
-            :math:`(B, C, H, W)`.
+        tensor (torch.Tensor): image of the form :math:`(*, H, W)`.
```
Comment (edgarriba) [inline kornia/utils/image.py:50]:
```suggestion
        tensor: image of the form :math:`(*, H, W)`.
```
typing is automatically generated now (don't fix all) I'll do the missing in separate PR

File: kornia/utils/image.py [hunk 5]
```diff
@@ -65,23 +64,25 @@ def _to_bchw(tensor: torch.Tensor) -> torch.Tensor:
...
     return tensor
 
 
 def _to_bcdhw(tensor: torch.Tensor) -> torch.Tensor:
     """Converts a PyTorch tensor image to BCDHW format.
 
     Args:
-        tensor (torch.Tensor): image of the form :math:`(D, H, W)`, :math:`(C, D, H, W)`, :math:`(D, H, W, C)` or
-            :math:`(B, C, D, H, W)`.
+        tensor (torch.Tensor): image of the form :math:`(*, D, H, W)`.
```
Comment (edgarriba) [inline kornia/utils/image.py:77]:
```suggestion
        tensor: image of the form :math:`(*, D, H, W)`.
```

File: kornia/utils/image.py [hunk 6]
```diff
@@ -154,22 +158,43 @@ def forward(self, x: np.ndarray) -> torch.Tensor:
...
 
 
-def perform_keep_shape(f):
+def perform_keep_shape_image(f):
     """TODO: where can we put this?"""
 
     @wraps(f)
     def _wrapper(input, *args, **kwargs):
         input_shape = input.shape
+        input = _to_bchw(input)
```
Comment (edgarriba) [inline kornia/utils/image.py:167]: add few high-level comments about what's happening within the functions

File: kornia/utils/image.py [hunk 7]
```diff
@@ -154,22 +158,43 @@ def forward(self, x: np.ndarray) -> torch.Tensor:
         return image_to_tensor(x, keepdim=self.keepdim)
 
 
-def perform_keep_shape(f):
+def perform_keep_shape_image(f):
```
Comment (edgarriba) [inline kornia/utils/image.py:161]:
```suggestion
def perform_keep_shape_image(f: Callable) -> torch.Tensor:
```
add typing wherever makes sense to improve code readability

File: test/enhance/test_adjust.py [hunk 8]
```diff
@@ -466,12 +466,12 @@ def test_gradcheck(self, device, dtype):
 
 class TestEqualize:
     def test_shape_equalize(self, device, dtype):
-        bs, channels, height, width = 1, 3, 4, 5
+        channels, height, width = 3, 4, 5
```
Comment (edgarriba) [inline test/enhance/test_adjust.py:469]: can you add `@pytest.mark.parametrize` and test different shapes ?

Comment (justanhduc) [reply test/enhance/test_adjust.py:469]: @edgarriba sure thing.

File: kornia/enhance/adjust.py (hunk 2 above)
Comment (justanhduc) [reply kornia/enhance/adjust.py:195]: IMO, `CxHxW` should be indicated when we explicitly need the channel dimension (for e.g., the color module), and `HxW` when we don't. I changed here cuz `(*, N)` doesn't really make sense for image, and users may think they need to flatten the image to use the function.

**Comment by justanhduc** [comment]
@edgarriba I rebased to master but there is diverged history with the current branch. Should I force push?

**Comment by edgarriba** [comment]
> @edgarriba I rebased to master but there is diverged history with the current branch. Should I force push?

make sure it doesn't conflict with any other thing

**Comment by justanhduc** [comment]
@edgarriba I have finalized this PR. Could you plz check?

**Review by edgarriba — APPROVED** [review]

**Comment by edgarriba** [comment]
@shijianjian @ducha-aiki @lferraz @Borda can you review/merge ?

**Comment by edgarriba** [comment]
@justanhduc something went wrong with rebasing. Could you try tu update so that we can merge ?

**Comment by justanhduc** [comment]
@edgarriba I rebased my commits.

**Review by shijianjian — APPROVED** [review]

**Comment by shijianjian** [comment]
@edgarriba Anything wrong with cuda? It has been queuing for more than 10 hours.

**Comment by justanhduc** [comment]
@edgarriba @shijianjian should I do anything to trigger the tests again?

**Comment by shijianjian** [comment]
> @edgarriba @shijianjian should I do anything to trigger the tests again?

No need. I will test it locally tonight and merge if it is alright. It might be something wrong on our side. We will figure it out later.

**Comment by shijianjian** [comment]
Hi @justanhduc, I got two failed tests (param gen for ColorJitter), can you fix it? You may reproduce the error on colab with:
```bash
!git clone https://github.com/justanhduc/kornia.git
!cd kornia && git checkout arbitrary-leading-dimensions
!cd kornia && python -m pytest test/augmentation/test_random_generator.py --dtype float64,float32 --device cuda
```

**Comment by justanhduc** [comment]
@shijianjian @edgarriba Thanks for noticing me. I can reproduce it in colab, but unfortunately, I couldn't on my local computer. This error is created by `order=torch.randperm(4, device=_device, dtype=_dtype).long()` which is controlled by setting seed. On my local computer, it returns the expected value, but somehow in colab, it doesn't. It's quite weird because the seed indeed controls all the generated params, not only `order`, and yet, only `order` is wrongly generated. Finally, I tried to run the tests at master, and it got the same error too. You can reproduce it with 
```
!git clone https://github.com/kornia/kornia.git
!cd kornia
!cd kornia && python -m pytest test/augmentation/test_random_generator.py --dtype float64,float32 --device cuda
```
I think the error is specific to colab itself. Please let me know if I can do anything related to this error.

**Comment by shijianjian** [comment]
> @shijianjian @edgarriba Thanks for noticing me. I can reproduce it in colab, but unfortunately, I couldn't on my local computer. This error is created by `order=torch.randperm(4, device=_device, dtype=_dtype).long()` which is controlled by setting seed. On my local computer, it returns the expected value, but somehow in colab, it doesn't. It's quite weird because the seed indeed controls all the generated params, not only `order`, and yet, only `order` is wrongly generated. Finally, I tried to run the tests at master, and it got the same error too. You can reproduce it with 
> 
> ```
> 
> !git clone https://github.com/kornia/kornia.git
> 
> !cd kornia
> 
> !cd kornia && python -m pytest test/augmentation/test_random_generator.py --dtype float64,float32 --device cuda
> 
> ```
> 
> I think the error is specific to colab itself. Please let me know if I can do anything related to this error.

Hi @justanhduc, I am now away from my computer. Can you please check if previous of Kornia worked or not? I think it is mergeable if this error is not introduced by this PR but by Colab.

**Comment by justanhduc** [comment]
Hi @shijianjian. Following your suggestion, I ran the test in colab and confirmed that all the recent `0.5.x` releases failed these two tests.

**Comment by shijianjian** [comment]
@justanhduc, Thanks. Can you raise a proper issue then we will investigate it later?

**Comment by justanhduc** [comment]
@shijianjian sure. Let me make an issue ticket.

**Pull request merged**